File: scrudbeans-annotation-processor/src/main/java/com/github/manosbatsis/scrudbeans/processor/ScrudModelAnnotationProcessor.java
package com.github.manosbatsis.scrudbeans.processor;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Filer;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.annotation.processing.SupportedSourceVersion;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.Name;
import javax.lang.model.element.TypeElement;
import javax.persistence.Entity;

import com.github.manosbatsis.scrudbeans.api.DtoMapper;
import com.github.manosbatsis.scrudbeans.api.mdd.ScrudModelProcessorException;
import com.github.manosbatsis.scrudbeans.api.mdd.annotation.model.ScrudBean;
import com.github.manosbatsis.scrudbeans.api.mdd.model.EntityModelDescriptor;
import com.github.manosbatsis.scrudbeans.api.mdd.model.ModelDescriptor;
import com.github.manosbatsis.scrudbeans.api.mdd.model.ScrudModelDescriptor;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.TypeSpec;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Annotation processor that generates SCRUD components
 * for model annotated with @{@link ScrudBean}
 * and JPA specification predicate factories for models
 * annotated with @{@link Entity}
 */
@SupportedAnnotationTypes({
		"com.github.manosbatsis.scrudbeans.api.mdd.annotation.model.ScrudBean",
		"javax.persistence.Entity"
})
@SupportedSourceVersion(SourceVersion.RELEASE_8)
public class ScrudModelAnnotationProcessor extends AbstractProcessor {

	private static final Logger log = LoggerFactory.getLogger(ScrudModelAnnotationProcessor.class);

	private boolean complete = false;

	private Filer filer;

	/**
	 * {@inheritDoc}
	 */
	public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
		// short-circuit if there are multiple rounds
		if (complete) {
			log.info("Processor has already been executed, ignoring");
			return true;
		}
		log.info("ScrudModelAnnotationProcessor processing started");
		this.filer = processingEnv.getFiler();

		// Create JPA query predicate factories for each entity in the source path
		generateEntityPredicateFactories(roundEnv);
		// Create other SCRUD components for each model annotated with ScrudBean
		generateScrudComponents(roundEnv);
		// Claiming that annotations have been processed by this processor

		complete = true;
		return true;
	}

	/**
	 * Create SCRUD components for the target model
	 * @param roundEnv The current compilation round environment
	 */
	private void generateScrudComponents(RoundEnvironment roundEnv) {
		Set<? extends Element> annotatedModels = roundEnv.getElementsAnnotatedWith(ScrudBean.class);
		Map<Name, ScrudModelDescriptor> modelDescriptors = new HashMap<>();
		log.info("ScrudModelAnnotationProcessor found {} annotated classes", (annotatedModels != null ? annotatedModels.size() : 0));
		if (annotatedModels != null) {
			for (final Element element : annotatedModels) {
				try {
					if (element instanceof TypeElement) {
						final TypeElement typeElement = (TypeElement) element;
						// Parse model to something more convenient
						ScrudModelDescriptor descriptor = new ScrudModelDescriptor(processingEnv, typeElement);
						// Generate components for model
						generateDtoMappers(descriptor);
						createRepository(descriptor);
						createService(descriptor);
						createController(descriptor);
					}
					else {
						log.warn("Not an instance of TypeElement but annotated with ScrudBean: {}", element.getSimpleName());
					}
				}
				catch (ScrudModelProcessorException e) {
					e.printStackTrace();
				}
			}
		}
	}

	/**
	 * Create JPA query predicate factories for each entity in the source path
	 * @param roundEnv The current compilation round environment
	 */
	private void generateEntityPredicateFactories(RoundEnvironment roundEnv) {
		Set<? extends Element> entities = roundEnv.getElementsAnnotatedWith(Entity.class);
		for (final Element element : entities) {
			try {
				if (element instanceof TypeElement) {
					final TypeElement typeElement = (TypeElement) element;
					EntityModelDescriptor descriptor = new EntityModelDescriptor(processingEnv, typeElement);
					createPredicateFactory(descriptor);
				}
				else {
					log.warn("Not an instance of TypeElement but annotated with ScrudBean: {}", element.getSimpleName());
				}
			}
			catch (ScrudModelProcessorException e) {
				log.error("Error generating components for {}: " + e.getMessage(),
						element.getSimpleName(), e);
			}
		}
	}

	/**
	 * Create a SCRUD REST controller source file
	 * @param descriptor The target model descriptor 
	 * @return the written file
	 */
	private JavaFile createController(ScrudModelDescriptor descriptor) {
		TypeSpec typeSpec = TypeSpecBuilder.createController(descriptor);
		return writeJavaFile(descriptor, typeSpec, descriptor.getParentPackageName() + ".controller");
	}

	/**
	 * Create {@link DtoMapper}s for the ScudBeans' target DTOs
	 * @param descriptor The target model descriptor
	 * @return the mapper files
	 */
	private List<JavaFile> generateDtoMappers(ScrudModelDescriptor descriptor) {
		List<JavaFile> files = new LinkedList<>();
		descriptor.getDtoTypes().forEach((dtoClass) -> {
			TypeSpec typeSpec = TypeSpecBuilder.createDtoMapper(descriptor, dtoClass);
			files.add(writeJavaFile(
					descriptor,
					typeSpec,
					descriptor.getParentPackageName() + ".mapper"));
		});
		return files;
	}
	/**
	 * Create SCRUD service source files
	 * @param descriptor The target model descriptor
	 * @return the written files: interface and implementation
	 */
	private List<JavaFile> createService(ScrudModelDescriptor descriptor) {
		List<JavaFile> files = new LinkedList<>();
		// Ensure a service has not already been created
		String serviceQualifiedName = descriptor.getParentPackageName() +
				".service." + descriptor.getSimpleName() + "Service";
		TypeElement existing = processingEnv.getElementUtils().getTypeElement(serviceQualifiedName);
		if (Objects.isNull(existing)) {
			files.add(createServiceInterface(descriptor));
			files.add(createServiceImpl(descriptor));
		}
		else {
			log.debug("createService: {} already exists, skipping", serviceQualifiedName);
		}
		return files;
	}

	/**
	 * Create a SCRUD service interface source file
	 * @param descriptor The target model descriptor 
	 * @return the written file
	 */
	private JavaFile createServiceInterface(ScrudModelDescriptor descriptor) {
		TypeSpec typeSpec = TypeSpecBuilder.createServiceInterface(descriptor);
		return writeJavaFile(descriptor, typeSpec, descriptor.getParentPackageName() + ".service");

	}

	/**
	 * Create a SCRUD service implementation source file
	 * @param descriptor The target model descriptor 
	 * @return the written file
	 */
	private JavaFile createServiceImpl(ScrudModelDescriptor descriptor) {
		TypeSpec typeSpec = TypeSpecBuilder.createServiceImpl(descriptor);
		return writeJavaFile(descriptor, typeSpec, descriptor.getParentPackageName() + ".service");
	}

	/**
	 * Create a SCRUD repository source file
	 * @param descriptor The target model descriptor 
	 * @return the written file
	 */
	private JavaFile createRepository(ScrudModelDescriptor descriptor) {
		TypeSpec typeSpec = TypeSpecBuilder.createRepository(descriptor);
		return writeJavaFile(descriptor, typeSpec, descriptor.getParentPackageName() + ".repository");
	}

	/**
	 * Create a JPA specification predicate factory source file
	 * @param descriptor The target model descriptor 
	 * @return the written file
	 */
	private JavaFile createPredicateFactory(EntityModelDescriptor descriptor) {
		TypeSpec typeSpec = TypeSpecBuilder.createPredicateFactory(descriptor);
		return writeJavaFile(descriptor, typeSpec, descriptor.getParentPackageName() + ".specification");
	}

	/**
	 * Write and return a source file for the given {@link TypeSpec}
	 * @param typeSpec The target model type spec
	 * @param descriptor The target model descriptor
	 * @param packageName The target source file package
	 * @return the written file
	 */
	private JavaFile writeJavaFile(ModelDescriptor descriptor, TypeSpec typeSpec, String packageName) {
		JavaFile file = null;
		String fileObjectName = packageName + "." + typeSpec.name;
		try {
			TypeElement existing = processingEnv.getElementUtils().getTypeElement(fileObjectName);
			if (existing == null) {
				log.debug("writeJavaFile for {}, filer: {}", fileObjectName, filer);
				file = JavaFile
						.builder(packageName, typeSpec)
						.build();
				file.writeTo(filer);
			}
			else {
				log.debug("writeJavaFile: Skipping for {} as it already exists", fileObjectName);
			}
		}
		catch (Exception e) {
			log.error("writeJavaFile: Error creating file for {}: " + e.getMessage(), fileObjectName, e);
		}
		return file;
	}

}

File: scrudbeans-annotation-processor-kotlin/src/main/kotlin/com/github/manosbatsis/scrudbeans/processor/kotlin/ScrudModelAnnotationProcessor.kt
package com.github.manosbatsis.scrudbeans.processor.kotlin

import com.github.manosbatsis.kotlin.utils.ProcessingEnvironmentAware
import com.github.manosbatsis.scrudbeans.api.annotation.model.ScrudBean
import com.github.manosbatsis.scrudbeans.api.mdd.model.IdentifierAdapter
import com.github.manosbatsis.scrudbeans.processor.kotlin.descriptor.ModelDescriptor
import com.github.manosbatsis.scrudbeans.processor.kotlin.descriptor.ScrudModelDescriptor
import com.squareup.kotlinpoet.*
import jakarta.persistence.Entity
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.util.LinkedList
import java.util.Objects
import java.util.Properties
import javax.annotation.processing.*
import javax.lang.model.SourceVersion
import javax.lang.model.element.Name
import javax.lang.model.element.TypeElement
import javax.tools.Diagnostic
import javax.tools.StandardLocation

/**
 * Annotation processor that generates SCRUD components
 * for model annotated with @[ScrudBean]
 * and JPA specification predicate factories for models
 * annotated with @[Entity]
 */
@SupportedAnnotationTypes("com.github.manosbatsis.scrudbeans.api.annotation.model.ScrudBean")
@SupportedSourceVersion(SourceVersion.RELEASE_17)
class ScrudModelAnnotationProcessor : AbstractProcessor(), ProcessingEnvironmentAware {

    companion object {
        private val log = LoggerFactory.getLogger(ScrudModelAnnotationProcessor::class.java)
        const val BLOCK_FUN_NAME = "block"
        const val KAPT_KOTLIN_SCRUDBEANS_GENERATED_OPTION_NAME = "kapt.kotlin.vaultaire.generated"
        const val KAPT_KOTLIN_GENERATED_OPTION_NAME = "kapt.kotlin.generated"

        val typeParameterStar = WildcardTypeName.producerOf(Any::class.asTypeName().copy(nullable = true))
    }

    private var complete = false
    private val typeSpecBuilder by lazy { TypeSpecBuilder(processingEnv) }
    private lateinit var filer: Filer

    // Config properties, i.e. "application.properties" from the classpath
    private lateinit var configProps: Properties

    /** Implement [ProcessingEnvironmentAware.processingEnvironment] for access to a [ProcessingEnvironment] */
    override val processingEnvironment by lazy {
        processingEnv
    }

    val sourceRootFile by lazy {
        val sourceRootFile = File(generatedSourcesRoot)
        sourceRootFile.mkdir()
        sourceRootFile
    }
    val generatedSourcesRoot: String by lazy {
        processingEnv.options[KAPT_KOTLIN_SCRUDBEANS_GENERATED_OPTION_NAME]
            ?: processingEnv.options[KAPT_KOTLIN_GENERATED_OPTION_NAME]
            ?: defaultGeneratedSourcesDir()
    }

    private fun defaultGeneratedSourcesDir(): String {
        val path = filer.createSourceFile("ScrudBeansGeneratedSourcesRoot")
            .let { File(it.toUri().toURL().file).parent }
        val msg = "No kapt.kotlin.generated option provided. Using $path"
        this.processingEnv.messager.printMessage(Diagnostic.Kind.MANDATORY_WARNING, msg + "\n")
        return path
    }

    /**
     * {@inheritDoc}
     */
    override fun process(annotations: Set<TypeElement>, roundEnv: RoundEnvironment): Boolean {
        // short-circuit if there are multiple rounds
        if (complete) {
            processingEnv.noteMessage { "Processor has already been executed, ignoring" }
            return true
        }
        processingEnv.noteMessage { "ScrudModelAnnotationProcessor processing started" }
        // Init a filer
        this.filer = processingEnv.filer
        // Load config/properties
        configProps = this.loadProperties()
        // Create JPA query predicate factories for each entity in the source path
        // generateEntityPredicateFactories(roundEnv)
        // Create other SCRUD components for each model annotated with ScrudBean
        generateScrudComponents(roundEnv)
        // Claiming that annotations have been processed by this processor
        complete = true
        return true
    }

    /**
     * Create SCRUD components for the target model
     * @param roundEnv The current compilation round environment
     */
    private fun generateScrudComponents(roundEnv: RoundEnvironment) {
        val annotatedModels = roundEnv.getElementsAnnotatedWith(ScrudBean::class.java)
        val modelDescriptors = HashMap<Name, ScrudModelDescriptor>()
        processingEnv.noteMessage { "ScrudModelAnnotationProcessor found ${annotatedModels?.size ?: 0} annotated classes" }
        if (annotatedModels != null) {
            for (element in annotatedModels) {
                try {
                    if (element is TypeElement) {
                        processingEnv.noteMessage { "ScrudModelAnnotationProcessor processing ${element.simpleName}" }
                        // Parse model to something more convenient
                        val descriptor = ScrudModelDescriptor(processingEnv, element, configProps)
                        // Mappers for manual DTOs
                        // generateDtoMappers(descriptor)
                        generateDto(descriptor)
                        createIdAdapters(descriptor)
                        createRepository(descriptor)
                        createService(descriptor)
                        createController(descriptor)
                    } else {
                        element.errorMessage { "Not an instance of TypeElement but annotated with ScrudBean: ${element.simpleName}" }
                    }
                } catch (e: Throwable) {
                    element.errorMessage { "Failed processing ScrudBean annotation for ${element.simpleName}: ${e.message ?: e.cause?.message}" }
                    e.printStackTrace()
                    throw e
                }
            }
        }
    }

    /**
     * Create a DTO source file
     * @param descriptor The target model descriptor
     * @return the written file
     */
    private fun generateDto(descriptor: ScrudModelDescriptor): FileSpec? {
        val typeSpec = typeSpecBuilder.dtoSpecBuilder(descriptor, sourceRootFile)
        return writeKotlinFile(descriptor, typeSpec, descriptor.packageName)
    }

    /**
     * Create a SCRUD REST controller source file
     * @param descriptor The target model descriptor
     * @return the written file
     */
    private fun createController(descriptor: ScrudModelDescriptor): FileSpec? {
        return if (ScrudBean.NONE != descriptor.scrudBean.controllerSuperClass) {
            writeKotlinFile(
                descriptor,
                typeSpecBuilder.createController(descriptor),
                descriptor.parentPackageName + ".controller",
            )
        } else {
            null
        }
    }

    /**
     * Create SCRUD service source files
     * @param descriptor The target model descriptor
     * @return the written files: interface and implementation
     */
    private fun createService(descriptor: ScrudModelDescriptor): List<FileSpec?> {
        val files = LinkedList<FileSpec?>()
        // Ensure a service has not already been created
        val serviceQualifiedName = descriptor.parentPackageName + ".service." + descriptor.simpleName + "Service"
        val existing = processingEnv.elementUtils.getTypeElement(serviceQualifiedName)
        if (Objects.isNull(existing)) {
            files.add(createServiceInterface(descriptor))
            files.add(createServiceImpl(descriptor))
        } else {
            processingEnv.noteMessage { "createService: $serviceQualifiedName} already exists, skipping" }
        }
        return files
    }

    /**
     * Create a SCRUD service interface source file
     * @param descriptor The target model descriptor
     * @return the written file
     */
    private fun createServiceInterface(descriptor: ScrudModelDescriptor): FileSpec? {
        val typeSpec = typeSpecBuilder.createServiceInterface(descriptor)
        return writeKotlinFile(descriptor, typeSpec, descriptor.parentPackageName + ".service")
    }

    /**
     * Create a SCRUD service implementation source file
     * @param descriptor The target model descriptor
     * @return the written file
     */
    private fun createServiceImpl(descriptor: ScrudModelDescriptor): FileSpec? {
        val typeSpec = typeSpecBuilder.createServiceImpl(descriptor)
        return writeKotlinFile(descriptor, typeSpec, descriptor.parentPackageName + ".service")
    }

    /**
     * Create a SCRUD repository source file
     * @param descriptor The target model descriptor
     * @return the written file
     */
    private fun createRepository(descriptor: ScrudModelDescriptor): FileSpec? {
        val typeSpec = typeSpecBuilder.createRepository(descriptor)
        return writeKotlinFile(descriptor, typeSpec, descriptor.parentPackageName + ".repository")
    }

    /**
     * Create an [IdentifierAdapter] implementation
     * @param descriptor The target model descriptor
     * @return the written file
     */
    private fun createIdAdapters(descriptor: ScrudModelDescriptor): List<FileSpec> {
        return listOf(
            ClassName(descriptor.packageName, descriptor.simpleName),
        )
            .mapNotNull {
                writeKotlinFile(descriptor, typeSpecBuilder.createIdAdapter(it, descriptor), descriptor.packageName)
            }
    }

    /**
     * Write and return a source file for the given [TypeSpec]
     * @param typeSpec The target model type spec
     * @param descriptor The target model descriptor
     * @param packageName The target source file package
     * @return the written file
     */
    private fun writeKotlinFile(descriptor: ModelDescriptor, typeSpec: TypeSpec, packageName: String): FileSpec? {
        val fileObjectName = packageName + "." + typeSpec.name!!
        var file: FileSpec? = null
        try {
            val existing = processingEnv.elementUtils.getTypeElement(fileObjectName)
            if (existing == null) {
                processingEnv.noteMessage { "writeJavaFile for $fileObjectName" }
                file = FileSpec.builder(packageName, typeSpec.name!!)
                    .addFileComment("-------------------- DO NOT EDIT -------------------\n")
                    .addFileComment(" This file is automatically generated by scrudbeans,\n")
                    .addFileComment(" see https://manosbatsis.github.io/scrudbeans\n")
                    .addFileComment(" To edit this file, copy it to the appropriate package \n")
                    .addFileComment(" in your src/main/kotlin folder and edit there. \n")
                    .addFileComment("----------------------------------------------------")
                    .addType(typeSpec)
                    .build()
                file.writeTo(sourceRootFile)
            } else {
                processingEnv.noteMessage { "writeJavaFile: Skipping for $fileObjectName as it already exists" }
            }
        } catch (e: Exception) {
            processingEnv.noteMessage { "writeJavaFile: Error creating file for $fileObjectName: ${e.message}" }
            throw e
        }

        return file
    }

    private fun loadProperties(): Properties {
        var props = Properties()
        try {
            val fileObject = this.filer!!
                .getResource(StandardLocation.CLASS_OUTPUT, "", "application.properties")
            props.load(fileObject.openInputStream())
            processingEnv.noteMessage { "loadProperties, props: $props" }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return props
    }
}
